web/views.py

```
@@ -3101,3 +3105,47 @@ def get_object(self):
+            return redirect("certificate_list")
+    else:
+        form = CertificateForm()
+    
+    return render(request, "certificates/upload.html", {"form": form})
+
+def certificate_detail(request, uuid):
+    certificate = get_object_or_404(Certificate, uuid=uuid)
+    
+    # Only allow viewing if certificate is public or user is the owner
+    if not certificate.is_public and (not request.user.is_authenticated or request.user != certificate.user):
+        raise Http404("Certificate not found")
```

Ruff (0.8.2)

Undefined name `Http404`

(F821)